### PR TITLE
[Core] Fix node affinity strategy when resource is empty (#25344)

### PR DESCRIPTION
**Pull request by scv119**
Why are these changes needed?
Today, Ray scheduler always pick a random node if the resource requirement is empty, regardless of scheduling policy/strategy.

However, for node affinity scheduling policy, we should not pick random policy but try to stick to the node affinity constraints.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
thanks

**Review by zhe-thoughts — APPROVED** [review]

**Comment by scv119** [comment]
Will wait for CI before merge it.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:115]: Shouldn't we also consider soft case?

**Comment by krfricke** [comment]
Can you open the PR for the `releases/1.13.0` branch?

**Pull request closed without merging**
